builtin/k8s/platform.go

```diff
@@ -657,47 +819,91 @@ func (p *Platform) Status(
 	sg := ui.StepGroup()
 	defer sg.Wait()
 
-	step := sg.Add("Gathering health report for Kubernetes platform...")
+	step := sg.Add("Gathering health report for Kubernetes deployment...")
 	defer func() { step.Abort() }() // Defer in func in case more steps are added to this func in the future
 
-	csInfo, err := p.getClientset()
-	if err != nil {
-		return nil, err
-	}
-	clientSet := csInfo.Clientset
-	namespace := csInfo.Namespace
-	if p.config.Namespace != "" {
-		namespace = p.config.Namespace
+	rm := p.resourceManager(log, nil)
+
+	// If we don't have resource state, this state is from an older version
+	// and we need to manually recreate it.
+	if deployment.ResourceState == nil {
+		rm.Resource("deployment").SetState(&Resource_Deployment{
+			Name: deployment.Name,
+		})
+	} else {
+		// Load our set state
+		if err := rm.LoadState(deployment.ResourceState); err != nil {
+			return nil, err
+		}
 	}
 
-	podClient := clientSet.CoreV1().Pods(namespace)
-	podLabelId := fmt.Sprintf("app=%s", deployment.Name)
-	podList, err := podClient.List(ctx, metav1.ListOptions{LabelSelector: podLabelId})
+	resources, err := rm.StatusAll(ctx, log, sg, ui)
 	if err != nil {
-		ui.Output(
-			"Error listing pods to determine application health: %s", clierrors.Humanize(err),
-			terminal.WithErrorStyle(),
-		)
-		return nil, err
+		return nil, status.Errorf(codes.Internal, "resource manager failed to generate resource statuses: %s", err)
 	}
 
+	// NOTE(briancain): Replace ui.Status with StepGroups once this bug
+	// has been fixed: https://github.com/hashicorp/waypoint/issues/1536
+	st := ui.Status()
+	defer st.Close()
+	st.Update("Determining overall container health...")
+
 	// Create our status report
-	step.Update("Building status report for running pods...")
-	result := buildStatusReport(podList)
+	result := sdk.StatusReport{
+		External:      true,
+		GeneratedTime: timestamppb.Now(),
+		Resources:     resources,
+	}
+
+	// Figure out our overall status based on our resource list. If all pods are in the same state, the
+	// status report state will match. If pods are in mixes states, we'll return a PARTIAL health
+	// and a detailed health message.
+
+	// Create a map of how many pods are in each status
+	healths := make(map[sdk.StatusReport_Health]int)
+	for _, r := range resources {
+		if r.Type != "pod" {
+			// We only use pod healths to determine overall status
+			continue
+		}
+		if _, ok := healths[r.Health]; !ok {
+			healths[r.Health] = 1
+		} else {
+			healths[r.Health]++
+		}
```

Comment:
Nit:

```suggestion
		healths[r.Health]++
```

Reply:
https://play.golang.com/p/Vt-NrbUhPyF

Reply:
Would you look at that! 👍

Reply:
Zero value of a non-existent key is a zero value. 😄